File: java/base/src/main/vsdk/toolkit/common/linealAlgebra/Vector4Df.java
package vsdk.toolkit.common.linealAlgebra;

import java.io.Serial;
import vsdk.toolkit.common.FundamentalEntity;
import vsdk.toolkit.common.VSDK;

public final class Vector4Df extends FundamentalEntity {
    @Serial
    private static final long serialVersionUID = 20260514L;

    private final float x;
    private final float y;
    private final float z;
    private final float w;

    public Vector4Df(float x, float y, float z, float w) {
        this.x = x;
        this.y = y;
        this.z = z;
        this.w = w;
    }

    public Vector4Df(Vector4Df other) {
        this(other.x, other.y, other.z, other.w);
    }

    public Vector4Df(Vector4Dd other) {
        this((float) other.x(), (float) other.y(), (float) other.z(), (float) other.w());
    }

    public Vector4Df(Vector3Df other) {
        this(other.x(), other.y(), other.z(), 1f);
    }

    public Vector4Df multiply(float a) {
        return new Vector4Df(a * x, a * y, a * z, a * w);
    }

    public Vector4Df dividedByW() {
        if ( Math.abs(w) < (float) VSDK.EPSILON ) return this;
        return new Vector4Df(x / w, y / w, z / w, 1f);
    }

    public float length() {
        return (float) Math.sqrt(x * x + y * y + z * z + w * w);
    }

    public Vector4Df add(Vector4Df b) {
        return new Vector4Df(x + b.x, y + b.y, z + b.z, w + b.w);
    }

    public Vector4Df withX(float nx) { return new Vector4Df(nx, y, z, w); }
    public Vector4Df withY(float ny) { return new Vector4Df(x, ny, z, w); }
    public Vector4Df withZ(float nz) { return new Vector4Df(x, y, nz, w); }
    public Vector4Df withW(float nw) { return new Vector4Df(x, y, z, nw); }
    public float x() { return x; }
    public float y() { return y; }
    public float z() { return z; }
    public float w() { return w; }

    public boolean epsilonEquals(Vector4Df other) {
        return epsilonEquals(other, (float) VSDK.EPSILON);
    }

    public boolean epsilonEquals(Vector4Df other, float epsilon) {
        if ( other == null ) return false;
        if ( epsilon < 0f ) throw new IllegalArgumentException("epsilon must be >= 0");
        return Math.abs(x - other.x) <= epsilon &&
               Math.abs(y - other.y) <= epsilon &&
               Math.abs(z - other.z) <= epsilon &&
               Math.abs(w - other.w) <= epsilon;
    }
}

File: typescript/base/src/main/vsdk/toolkit/common/linealAlgebra/Vector4Df.ts
import { FundamentalEntity } from "../FundamentalEntity.js";
import { VSDK } from "../VSDK.js";
import { Vector3Df } from "./Vector3Df.js";
import { Vector4Dd } from "./Vector4Dd.js";
const f = Math.fround;
export class Vector4Df extends FundamentalEntity {
    private readonly xv: number;
    private readonly yv: number;
    private readonly zv: number;
    private readonly wv: number;
    public constructor(x: number, y: number, z: number, w: number);
    public constructor(o: Vector4Df | Vector4Dd | Vector3Df);
    public constructor(a: number | Vector4Df | Vector4Dd | Vector3Df, b = 0, c = 0, d = 1) {
        super();
        if (typeof a === "number") {
            this.xv = f(a);
            this.yv = f(b);
            this.zv = f(c);
            this.wv = f(d);
        } else if (a instanceof Vector3Df) {
            this.xv = a.x();
            this.yv = a.y();
            this.zv = a.z();
            this.wv = 1;
        } else {
            this.xv = f(a.x());
            this.yv = f(a.y());
            this.zv = f(a.z());
            this.wv = f(a.w());
        }
    }
    public multiply(a: number): Vector4Df {
        return new Vector4Df(f(a * this.xv), f(a * this.yv), f(a * this.zv), f(a * this.wv));
    }
    public dividedByW(): Vector4Df {
        return Math.abs(this.wv) < f(VSDK.EPSILON)
            ? this
            : new Vector4Df(f(this.xv / this.wv), f(this.yv / this.wv), f(this.zv / this.wv), 1);
    }
    public length(): number {
        return f(Math.hypot(this.xv, this.yv, this.zv, this.wv));
    }
    public add(o: Vector4Df): Vector4Df {
        return new Vector4Df(f(this.xv + o.xv), f(this.yv + o.yv), f(this.zv + o.zv), f(this.wv + o.wv));
    }
    public withX(x: number): Vector4Df {
        return new Vector4Df(x, this.yv, this.zv, this.wv);
    }
    public withY(y: number): Vector4Df {
        return new Vector4Df(this.xv, y, this.zv, this.wv);
    }
    public withZ(z: number): Vector4Df {
        return new Vector4Df(this.xv, this.yv, z, this.wv);
    }
    public withW(w: number): Vector4Df {
        return new Vector4Df(this.xv, this.yv, this.zv, w);
    }
    public x(): number {
        return this.xv;
    }
    public y(): number {
        return this.yv;
    }
    public z(): number {
        return this.zv;
    }
    public w(): number {
        return this.wv;
    }
    public epsilonEquals(o: Vector4Df | null, e = f(VSDK.EPSILON)): boolean {
        if (o === null) return false;
        if (e < 0) throw new RangeError("epsilon must be >= 0");
        return (
            Math.abs(this.xv - o.xv) <= e &&
            Math.abs(this.yv - o.yv) <= e &&
            Math.abs(this.zv - o.zv) <= e &&
            Math.abs(this.wv - o.wv) <= e
        );
    }
}
